src/sampler.jl

```diff
@@ -1,3 +1,55 @@
+##
+## Interface functions
+##
+
+function init(rng::AbstractRNG, h::Hamiltonian, θ::AbstractVector{T}) where {T<:Real}
```

Comment:
Perhaps change `init` into `sample_init`to to match `Interface.jl` API in https://github.com/TuringLang/Turing.jl/pull/793?

cc @cpfiffer Do we have immutable versions of `sample_init!` and `step!` etc? If not, are there any reasons?

Reply:
I guess this is more similar to `transition_init`?

Reply:
It feels like a sampler level initialization since it is only performed once for a chain. Maybe @cpfiffer can give a better answer here.

Reply:
We can actually remove this `init` by performing momentum sampling in the beginning of each transition.

Reply:
OK we can't, there is still `h = update(h, θ)`

Reply:
`init` is currently used in Turing in the `HMCState` constructor.

I think from the AHMC side or other sampler side, it would be nice to support an immutable way to intergate, right? Also I don't think there is a solution that a `step` implementation in AHMC can be directly used in Turing. We need to wrap it in some way so that things like `Sampler` and `State` works. Is it a correct understanding?

Reply:
> z (I think) contains the state information

Yes. But just to add that it doesn't have all the information needed for adapation. Adapation requires additionally the acceptane rate.

Reply:
So what do you suggest me to do now?

Reply:
By `transition` do you mean `step` or something?

Reply:
Now I get it.